### ISSUE TITLE
Mono support

### ISSUE THREAD

**Issue by SilverDragon135** [qmfrederik/remoteviewing#2]
Hi is this library mono compatible? Can i build cross platform vnc viewer?

**Comment by qmfrederik** [comment]
I haven't tried Mono recently, but the library is compatible with .NET Core. I'm going to use the .NET Core version to have it run on Linux in the not-so-distant future; and from experience: if it works on .NET Core, it'll work on Mono, too.

If you encounter issues when running on Mono, let me know, and it's likely something we'd consider fixing.

**Comment by SilverDragon135** [comment]
Great thank you.

**Comment by SilverDragon135** [comment]
Hi, so I tested it with kubuntu and I experienced a few problems.
1. In the remote viewing example is a bug with GDI, but it should be related to mono and I don't have time now to find the solution... I will look at that, if we find the solution to:
2. Server example successfully started
- Krdc cannot even connect, but it seems like krdc cannot accept only password protected vnc - don't care = not that important
- with connection from Realvnc I experienced bug in:
internal static VncPixelFormat Decode (byte [] buffer, int offset) where VncUtility.DecodeUInt16BE returned 0 all the time and  this triggers exception in BitsFromMax.
I would like to find a solution, but I cannot exactly find a reference for Decode call. I found only in VncClient. cs and this file is not related to Server example. Can you point me to place, where is Decode called?
PS: for all of bits (red, green, blue) is 0s returned from VncUtility. Decode... (seems like a problem with passed buffer)
Thank you

**Comment by SilverDragon135** [comment]
OK, i checked tightvnc and everything works as expected. Now i will try correct win.forms to work under linux enviroment.
Here is complete error in case of fail connection with realvnc: 
Unhandled Exception:
System.ArgumentException: Value does not fall within the expected range.
  at RemoteViewing.Vnc.VncPixelFormat.BitsFromMax (Int32 max) <0x40ed1fc0 + 0x0011f> in <filename unknown>:0 
  at RemoteViewing.Vnc.VncPixelFormat.Decode (System.Byte[] buffer, Int32 offset) <0x40ed1db0 + 0x000b3> in <filename unknown>:0 
  at RemoteViewing.Vnc.Server.VncServerSession.HandleSetPixelFormat () <0x40ed1d10 + 0x0003f> in <filename unknown>:0 
  at RemoteViewing.Vnc.Server.VncServerSession.ThreadMain () <0x40e920b0 + 0x0032f> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x7f36c1a5c3d0 + 0x00099> in <filename unknown>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state,serveSyncCtx) <0x7f36c1a5a990 + 0x00178> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, BooleancCtx) <0x7f36c1a5a960 + 0x00020> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x7f36x00059> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart () <0x7f36c1a5c590 + 0x0002e> in <filename unknown>:0

**Comment by qmfrederik** [comment]
Yes, I've seen that before - the `remoteviewing` client works when using TightVNC but crashes out when using RealVNC.

If memory serves me well, that's because RealVNC uses palette encoding and that's not supported yet. So in short, it means that instead of sending the full color for each pixel, RealVNC will send a bunch of colors first, and then reference that color by its index for each pixel.

It would be great if the library would at least error out with a clear error message if it detects an encoding which is not supported rather than just crashing like it does now.

On a similar note, I think the library also uses `memcpy` in some locations to copy buffers around (that was a performance improvement at the time). Not sure whether that will work on Mono, though.

**Comment by SilverDragon135** [comment]
Thank you for fast response. It is really helpful :) About the mono. The main problem is that example is written in windows forms and in linux environment it is not stable (1 of 10 execution i can see the form). Next i will try port Windows.Forms library a example to mono gtk (hope that helps). And if you like we can then merge it with remoteviewer.

**Comment by SilverDragon135** [comment]
Update: 
Client: I found out, that mono is trolling me hard. I was able to test RemoteViewing.Example with few changes. It seems like mono cannot handle .net textBoxes - I changed them to  richTextBoxes and client example works fine. (ofc. nono input available, but not my concern) If you want, I can post here full source... (btw. tested on win7 .net 4.5.1 = working).

Server: I experienced, that ServerExample can handle only one connection. All connection are refused even after disconnecting  and Server needs to be restarted.
